src/utils/routeHelpers.ts

```diff
@@ -0,0 +1,160 @@
+import bcrypt from 'bcrypt';
+import { User } from '@prisma/client';
+import prismaClient from '../../prismaClient';
+import {
+  UserWithoutId, UserWithoutPasswordHash, LoginData, CreateUserData,
+} from '../types';
+
+/**
+ * - narrows username type to string, checks if username is valid
+ * - throws Error if username already exists
+ * - throws Error if username not valid
+ * @param username this should be a valid username
+ * @returns resolved promise with username as value
+ */
+const parseUsername = async (username: unknown): Promise<string> => {
+  const MINIMUM_USERNAME_LENGTH = 4;
+  if (typeof username !== 'string' || username.length < MINIMUM_USERNAME_LENGTH) {
+    const error = new Error('Invalid Username');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  const existingUser: User | null = await prismaClient.user.findUnique({
+    where: {
+      user_name: username,
+    },
+  });
+
+  if (existingUser) {
+    const error = new Error('User already exists');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  return Promise.resolve(username);
+};
+
+/**
+ * - narrows type of password to string, generates password hash from the password
+ * - throws error if password is missing or invalid ie less than 4 characters
+ * @param password this should be a valid password
+ * @returns Promise resolved to password hash
+ */
+const getPasswordHash = async (password: unknown): Promise<string> => {
+  const MINIMUM_PASSWORD_LENGTH = 4;
+  if (typeof password !== 'string' || password.length < MINIMUM_PASSWORD_LENGTH) {
+    const error = new Error('Invalid Password');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  const SALT_ROUNDS = 10;
+  const passwordHash = await bcrypt.hash(password, SALT_ROUNDS);
+  return Promise.resolve(passwordHash);
+};
+
+/**
+ * - performs input validation and type narrowing for input to POST /user
+ * - if username, password present and of type string, then returns an object containing input data
+ * - else throws Error
+ * @param object request.body should contain username and password
+ * @returns Promise resolved to user input to POST /user
+ */
+const generateCreateUserData = async (object: unknown): Promise<CreateUserData> => {
+  if (!object || typeof object !== 'object') {
+    throw new Error('Incorrect or missing data');
+  }
+
+  if ('username' in object && 'password' in object
+      && typeof object.username === 'string'
+      && typeof object.password === 'string') {
+    const createUserData = {
+      username: object.username,
+      password: object.password,
+    };
```

Comment:
In that case it would also have `dob`, `bio`. We can do that later. We will need to write input validation for them and also a test-suite.

Honestly, this was the most tough part of using TS. `req.body` could contain anything. To check that it only contains the relevant fields, narrow it to the relevant types, and then perform input validation on every field. I am still not satisfied with the final result, but at least we have something which works.

If you want I can add a couple of more fields or all the fields.

Reply:
Hmm, you're right, I don't think it's essential to have all the fields now.